rubin_sim/sim_archive/sim_archive.py

```diff
@@ -931,3 +934,189 @@ def compile_sim_archive_metadata_cli(*args):
         compilation_resource = ResourcePath(compilation_uri)
 
     compilation_resource = compile_sim_metadata(archive_uri, compilation_resource, append=append)
+
+
+def find_latest_prenight_sim_for_nights(
+    first_day_obs: str | None = None,
+    last_day_obs: str | None = None,
+    tags: tuple[str] = ("ideal", "nominal"),
+    max_simulation_age: int = 2,
+    archive_uri: str = "s3://rubin:rubin-scheduler-prenight/opsim/",
+    compilation_uri: str = "s3://rubin:rubin-scheduler-prenight/opsim/compiled_metadata_cache.h5",
+) -> pd.DataFrame:
+    """Find the most recent prenight simulation that covers a night.
+
+    Parameters
+    ----------
+    first_day_obs : `str` or  `None`
+        The date of the evening for the first night for which to get
+        a simulation. If `None`, then the current date will be used.
+    last_day_obs : `str` or  `None`
+        The date of the evening for the last night for which to get
+        a simulation. If `None`, then the current date will be used.
+    tags : `tuple[str]`
+        A tuple of tags to filter simulations by.
+        Defaults to ``('ideal', 'nominal')``.
+    max_simulation_age : `int`
+        The maximum age of simulations to consider, in days.
+        Simulations older than ``max_simulation_age`` will not be considered.
+        Defaults to 2.
+    archive_uri : `str`
+        The URI of the archive from which to fetch the simulation.
+        Defaults to ``s3://rubin:rubin-scheduler-prenight/opsim/``.
+    compilation_uri : `str`
+        The URI of the compiled metadata HDF5 file for efficient querying.
+        Defaults to
+        ``s3://rubin:rubin-scheduler-prenight/opsim/compiled_metadata_cache.h5``.
+
+    Returns
+    -------
+    sim_metadata : `dict`
+        A dictionary with metadata for the simulation.
+    """
+
+    if first_day_obs is None:
+        first_day_obs = Time(Time.now().mjd - 0.5, format="mjd").iso[:10]
+    if last_day_obs is None:
+        last_day_obs = first_day_obs
+
+    sim_metadata = read_archived_sim_metadata(
+        archive_uri, num_nights=max_simulation_age, compilation_resource=compilation_uri
+    )
```

Comment:
is sim_metadata a pandas dataframe? would it make sense to make it one, so that the section below becomes a query of the dataframe?

Reply:
sim_metadata is a list of dicts. It can be trivially turned into a dataframe, but doing that doesn't really simplify things here. Some of the elements we are concerned with are lists or dicts. Yes, a cell in a pandas DataFrame can be a dict or a list, but dealing with them is awkward, so using a DataFrame ends up making the code harder to read.

Reply:
I've reworked it so it only tries using ResourcePath if db_con is either not a str, or it can be parsed with pythons URL parser to get a valid scheme. So, it will only mention ResourcePath if it actually needs it.